src/aero_proc/src/lib.rs

```diff
@@ -0,0 +1,291 @@
+use quote::quote;
+use syn::{punctuated::Punctuated, spanned::Spanned, Expr, FnArg, Lit, Pat, Type};
+
+enum ArgType {
+    Array(bool, usize),
+    Slice(bool),
+    Pointer(bool),
+    Reference(bool),
+    String,
+    Path,
+}
+
+#[proc_macro_attribute]
+pub fn syscall(_: TokenStream, item: TokenStream) -> TokenStream {
+    let parsed_fn = syn::parse_macro_input!(item as syn::ItemFn);
+
+    if let Some(constness) = parsed_fn.sig.constness {
+        Diagnostic::spanned(
+            constness.span().unwrap(),
+            Level::Error,
+            "syscall functions cannot be const",
+        )
+        .emit();
+    }
+
+    if let Some(asyncness) = parsed_fn.sig.asyncness {
+        Diagnostic::spanned(
+            asyncness.span().unwrap(),
+            Level::Error,
+            "syscall functions cannot be async",
+        )
+        .emit();
+    }
+
+    if let Some(unsafety) = parsed_fn.sig.unsafety {
+        Diagnostic::spanned(
+            unsafety.span().unwrap(),
+            Level::Error,
+            "syscall functions cannot be unsafe",
+        )
+        .emit();
+    }
+
+    if let Some(_) = parsed_fn.sig.generics.lt_token {
+        let lt_span = parsed_fn.sig.generics.lt_token.span().unwrap();
+        let gt_span = parsed_fn.sig.generics.gt_token.span().unwrap();
+
+        Diagnostic::spanned(
+            lt_span.join(gt_span).unwrap(),
+            Level::Error,
+            "syscall functions cannot have generic parameters",
+        )
+        .emit();
+    }
+
+    let attrs = &parsed_fn.attrs;
+    let vis = &parsed_fn.vis;
+    let name = &parsed_fn.sig.ident;
+    let orig_args = &parsed_fn.sig.inputs;
+    let processed_args = process_args(orig_args);
+    let call_args = process_call_args(orig_args);
+    let ret = &parsed_fn.sig.output;
+    let body = &parsed_fn.block;
+
+    let result = quote! {
+        #(#attrs)* #vis fn #name(#(#processed_args),*) #ret {
+            #(#attrs)* fn inner_syscall(#orig_args) #ret {
+                #body
+            }
+
+            inner_syscall(#(#call_args),*)
+        }
+    };
+
+    result.into()
+}
+
+fn determine_arg_type(typ: &Type) -> Option<ArgType> {
+    match typ {
+        Type::Reference(typ) => match typ.elem.as_ref() {
+            Type::Array(array) => match &array.len {
+                Expr::Lit(lit) => match &lit.lit {
+                    Lit::Int(lit) => match lit.base10_parse() {
+                        Ok(len) => Some(ArgType::Array(typ.mutability.is_some(), len)),
+                        Err(err) => {
+                            Diagnostic::spanned(
+                                lit.span().unwrap(),
+                                Level::Error,
+                                &format!("failed to parse array length: {}", err),
+                            )
+                            .emit();
+
+                            // If we can't parse the array length, we just give it an arbitrary length.
```

Comment:
fixed in https://github.com/Andy-Python-Programmer/aero/pull/65/commits/3621bc1168f65d1ac60e7b2c50693e29f707de44 since now we automagically get the length of the array :^)